typing/typecore.ml

```diff
@@ -2000,6 +2003,12 @@ struct
       | Texp_apply ({exp_desc = Texp_ident (_, _, vd)}, [_, Some arg])
         when is_ref vd ->
           Use.guard (expression env arg)
+      | Texp_apply (e, args) when
+          List.exists (function (_,None) -> true | _ -> false) args ->
+          (* an expression abstracted over a missing argument *)
+          let arg env (_, eo) = option expression env eo in
+          let ty = Use.join (list arg env args) (expression env e) in
+          Use.(join (discard ty) (delay ty))
```

Comment:
I don't remember the details well, but isn't `join (discard ty) (delay ty)` equal to `delay ty`?

I think it could be nice if you merged the two cases. If it is the case that `inspect (join a b)` is equal to `join (inspect a) (inspect b)`, you could write a single logic, with a test that applies either a `delay` or an `inspect` depending on whether the application is delayed by a missing argument.

Reply:
> isn't `join (discard ty) (delay ty)` equal to `delay ty`?

It's equal to `discard ty`, I think.

Thanks for the suggestions.  I've merged the two cases, and added a comment.

Reply:
I checked and indeed `discard`, which is an alias for `guard`, is less permissive than `delay`, so their union is `discard`. This is still a bit non-intuitive to me, but I guess that in a call-by-value setting it is natural that evaluating now (and then not using) demands more than evaluating later.